train sentiment analysis

## Description
I ran the TrainSentimentAnalysis class with arguments ("-g", "1", "-e", "1", "-m", "2") in TrainSentimentAnalysisTest class, then the program throws an exception.

### Expected Behavior
train well

### Error Message
09:13:26.483 [main] DEBUG ai.djl.repository.Resource - Preparing artifact: MXNet, ai.djl.mxnet:glove:0.0.1 {"dimensions":"50"}
09:13:26.484 [main] DEBUG ai.djl.repository.AbstractRepository - Files have been downloaded already: /Users/hs/.djl.ai/cache/repo/model/nlp/word_embedding/ai/djl/mxnet/glove/50/0.0.1
09:13:26.882 [main] DEBUG ai.djl.BaseModel - Try to load model from /Users/hs/.djl.ai/cache/repo/model/nlp/word_embedding/ai/djl/mxnet/glove/50/0.0.1/glove-0000.params
09:13:26.883 [main] DEBUG ai.djl.BaseModel - Loading saved model: glove parameter
09:13:27.004 [main] DEBUG ai.djl.BaseModel - DJL model loaded successfully
09:13:28.270 [main] DEBUG ai.djl.repository.Resource - Preparing artifact: BasicDataset, ai.djl.basicdataset:stanford-movie-review:1.0 {}
09:13:28.271 [main] DEBUG ai.djl.repository.AbstractRepository - Files have been downloaded already: /Users/hs/.djl.ai/cache/repo/dataset/nlp/ai/djl/basicdataset/stanford-movie-review/1.0
09:13:31.347 [main] DEBUG ai.djl.repository.Resource - Preparing artifact: BasicDataset, ai.djl.basicdataset:stanford-movie-review:1.0 {}
09:13:31.347 [main] DEBUG ai.djl.repository.AbstractRepository - Files have been downloaded already: /Users/hs/.djl.ai/cache/repo/dataset/nlp/ai/djl/basicdataset/stanford-movie-review/1.0
09:13:34.254 [main] INFO  a.d.t.l.LoggingTrainingListener - Training on: cpu().
09:13:34.255 [main] INFO  a.d.t.l.LoggingTrainingListener - Load MXNet Engine Version 1.7.0 in 0.046 ms.
09:13:34.314 [main] ERROR ai.djl.training.dataset.DataIterable - java.lang.IllegalArgumentException: The desired shape has too many dimensions
09:13:34.352 [main] ERROR c.tigerobo.javadl.train.PtSATrainer - train failed!
java.lang.IllegalStateException: Data loading failed
	at ai.djl.training.dataset.DataIterable.next(DataIterable.java:155)
	at ai.djl.training.dataset.DataIterable.next(DataIterable.java:41)
	at ai.djl.training.EasyTrain.fit(EasyTrain.java:44)
	at com.tigerobo.javadl.train.PtSATrainer.train(PtSATrainer.java:99)
	at com.tigerobo.javadl.PtSATest.train(PtSATest.java:26)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:69)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:33)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:220)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:53)
Caused by: java.util.concurrent.ExecutionException: java.lang.IllegalArgumentException: The desired shape has too many dimensions
	at java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.util.concurrent.FutureTask.get(FutureTask.java:192)
	at ai.djl.training.dataset.DataIterable.next(DataIterable.java:152)
	... 26 common frames omitted
Caused by: java.lang.IllegalArgumentException: The desired shape has too many dimensions
	at ai.djl.mxnet.engine.MxNDArray.repeatsToMatchShape(MxNDArray.java:1466)
	at ai.djl.mxnet.engine.MxNDArray.repeat(MxNDArray.java:1308)
	at ai.djl.translate.PaddingStackBatchifier.batchify(PaddingStackBatchifier.java:68)
	at ai.djl.training.dataset.DataIterable.fetch(DataIterable.java:175)
	at ai.djl.training.dataset.DataIterable.access$100(DataIterable.java:41)
	at ai.djl.training.dataset.DataIterable$PreFetchCallable.call(DataIterable.java:224)
	at ai.djl.training.dataset.DataIterable$PreFetchCallable.call(DataIterable.java:211)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

## How to Reproduce?
https://github.com/awslabs/djl/blob/v0.8.0/examples/src/main/java/ai/djl/examples/training/TrainSentimentAnalysis.java

### Steps to reproduce
(Paste the commands you ran that produced the error.)

1.set the arguments for main function as ("-g", "1", "-e", "1", "-m", "2")
2.run TrainSentimentAnalysis class

## What have you tried to solve it?

1.I tried to change the new shape's dimension on line 196, but I don't know set to set well

## Environment Info

Please run the command `./gradlew debugEnv` from the root directory of DJL (if necessary, clone DJL first). It will output information about your system, environment, and installation that can help us debug your issue. Paste the output of the command below:

[INFO ] - ----------System Properties----------
[INFO ] - io.netty.allocator.numHeapArenas: 1
[INFO ] - java.vendor.version: JBR-11.0.8.10-944.34-jcef
[INFO ] - java.specification.version: 1.8
[INFO ] - pty4j.preferred.native.folder: /Applications/IntelliJ IDEA CE.app/Contents/lib/pty4j-native
[INFO ] - file.encoding: UTF-8
[INFO ] - pty4j.tmpdir: /Users/hs/Library/Caches/JetBrains/IdeaIC2020.2/tmp
[INFO ] - jdk.module.illegalAccess.silent: true
[INFO ] - org.gradle.internal.publish.checksums.insecure: true
[INFO ] - java.vm.name: OpenJDK 64-Bit Server VM
[INFO ] - line.separator: 

[INFO ] - idea.jre.check: true
[INFO ] - user.dir: /Users/hs/Projects/niubi/djl/integration
[INFO ] - java.ext.dirs: /Users/hs/Library/Java/Extensions:/Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home/jre/lib/ext:/Library/Java/Extensions:/Network/Library/Java/Extensions:/System/Library/Java/Extensions:/usr/lib/java
[INFO ] - idea.max.intellisense.filesize: 2500
[INFO ] - os.arch: x86_64
[INFO ] - java.awt.graphicsenv: sun.awt.CGraphicsEnvironment
[INFO ] - kotlinx.coroutines.debug: off
[INFO ] - jna.platform.library.path: /usr/lib:/usr/lib
[INFO ] - java.rmi.server.hostname: 127.0.0.1
[INFO ] - java.vendor.url.bug: https://bugreport.java.com/bugreport/
[INFO ] - java.io.tmpdir: /var/folders/3g/2drk3qmj39b7h61rwf6w8f8w0000gn/T/
[INFO ] - sun.cpu.endian: little
[INFO ] - sun.os.patch.level: unknown
[INFO ] - io.netty.allocator.numDirectArenas: 1
[INFO ] - java.vm.specification.version: 1.8
[INFO ] - idea.popup.weight: heavy
[INFO ] - idea.xdebug.key: -Xdebug
[INFO ] - java.vm.specification.name: Java Virtual Machine Specification
[INFO ] - jdk.debug: release
[INFO ] - io.netty.machineId: 28:f0:76:ff:fe:16:65:0e
[INFO ] - sun.java.command: ai.djl.integration.util.DebugEnvironment
[INFO ] - java.class.path: /Users/hs/Projects/niubi/djl/integration/build/classes/java/main:/Users/hs/Projects/niubi/djl/integration/build/resources/main:/Users/hs/.gradle/caches/modules-2/files-2.1/commons-cli/commons-cli/1.4/c51c00206bb913cd8612b24abd9fa98ae89719b1/commons-cli-1.4.jar:/Users/hs/.gradle/caches/modules-2/files-2.1/org.apache.logging.log4j/log4j-slf4j-impl/2.13.3/7cca27a921a18645139cf651c04b83b1a19cfd76/log4j-slf4j-impl-2.13.3.jar:/Users/hs/Projects/niubi/djl/basicdataset/build/libs/basicdataset-0.8.0-SNAPSHOT.jar:/Users/hs/Projects/niubi/djl/model-zoo/build/libs/model-zoo-0.8.0-SNAPSHOT.jar:/Users/hs/Projects/niubi/djl/testing/build/libs/testing-0.8.0-SNAPSHOT.jar:/Users/hs/.gradle/caches/modules-2/files-2.1/org.testng/testng/7.1.0/7504709f2d0b6e5070d1c6a0cae6c1c605489f8/testng-7.1.0.jar:/Users/hs/Projects/niubi/djl/mxnet/mxnet-model-zoo/build/libs/mxnet-model-zoo-0.8.0-SNAPSHOT.jar:/Users/hs/Projects/niubi/djl/mxnet/mxnet-engine/build/libs/mxnet-engine-0.8.0-SNAPSHOT.jar:/Users/hs/.gradle/caches/modules-2/files-2.1/ai.djl.mxnet/mxnet-native-auto/1.7.0-backport/ee5b368ef94c1fcec4ade4a6edacffb420fefce7/mxnet-native-auto-1.7.0-backport.jar:/Users/hs/Projects/niubi/djl/api/build/libs/api-0.8.0-SNAPSHOT.jar:/Users/hs/.gradle/caches/modules-2/files-2.1/org.slf4j/slf4j-api/1.7.30/b5a4b6d16ab13e34a88fae84c35cd5d68cac922c/slf4j-api-1.7.30.jar:/Users/hs/.gradle/caches/modules-2/files-2.1/org.apache.logging.log4j/log4j-core/2.13.3/4e857439fc4fe974d212adaaaa3b118b8b50e3ec/log4j-core-2.13.3.jar:/Users/hs/.gradle/caches/modules-2/files-2.1/org.apache.logging.log4j/log4j-api/2.13.3/ec1508160b93d274b1add34419b897bae84c6ca9/log4j-api-2.13.3.jar:/Users/hs/.gradle/caches/modules-2/files-2.1/org.apache.commons/commons-csv/1.8/37ca9a9aa2d4be2599e55506a6d3170dd7a3df4/commons-csv-1.8.jar:/Users/hs/.gradle/caches/modules-2/files-2.1/com.beust/jcommander/1.72/6375e521c1e11d6563d4f25a07ce124ccf8cd171/jcommander-1.72.jar:/Users/hs/.gradle/caches/modules-2/files-2.1/com.google.inject/guice/4.1.0/faf9ee8ac09eafd1128091426dd367a8c0085d55/guice-4.1.0-no_aop.jar:/Users/hs/.gradle/caches/modules-2/files-2.1/org.yaml/snakeyaml/1.21/18775fdda48574784f40b47bf478ab0593f92e4d/snakeyaml-1.21.jar:/Users/hs/.gradle/caches/modules-2/files-2.1/com.google.code.gson/gson/2.8.6/9180733b7df8542621dc12e21e87557e8c99b8cb/gson-2.8.6.jar:/Users/hs/.gradle/caches/modules-2/files-2.1/net.java.dev.jna/jna/5.3.0/4654d1da02e4173ba7b64f7166378847db55448a/jna-5.3.0.jar:/Users/hs/.gradle/caches/modules-2/files-2.1/org.apache.commons/commons-compress/1.20/b8df472b31e1f17c232d2ad78ceb1c84e00c641b/commons-compress-1.20.jar:/Users/hs/.gradle/caches/modules-2/files-2.1/javax.inject/javax.inject/1/6975da39a7040257bd51d21a231b76c915872d38/javax.inject-1.jar:/Users/hs/.gradle/caches/modules-2/files-2.1/aopalliance/aopalliance/1.0/235ba8b489512805ac13a8f9ea77a1ca5ebe3e8/aopalliance-1.0.jar:/Users/hs/.gradle/caches/modules-2/files-2.1/com.google.guava/guava/19.0/6ce200f6b23222af3d8abb6b6459e6c44f4bb0e9/guava-19.0.jar
[INFO ] - java.runtime.name: OpenJDK Runtime Environment
[INFO ] - java.vm.vendor: AdoptOpenJDK
[INFO ] - gopherProxySet: false
[INFO ] - org.gradle.internal.http.socketTimeout: 120000
[INFO ] - library.jansi.path: /Users/hs/.gradle/native/jansi/1.18/osx
[INFO ] - java.library.path: /Users/hs/Library/Java/Extensions:/Library/Java/Extensions:/Network/Library/Java/Extensions:/System/Library/Java/Extensions:/usr/lib/java:.
[INFO ] - java.vendor.url: http://java.oracle.com/
[INFO ] - apple.laf.useScreenMenuBar: true
[INFO ] - java.specification.vendor: Oracle Corporation
[INFO ] - java.version.date: 2020-07-14
[INFO ] - user.home: /Users/hs
[INFO ] - idea.dynamic.classpath: false
[INFO ] - idea.platform.prefix: Idea
[INFO ] - log4j.defaultInitOverride: true
[INFO ] - io.netty.processId: 31859
[INFO ] - org.gradle.internal.http.connectionTimeout: 60000
[INFO ] - awt.toolkit: sun.lwawt.macosx.LWCToolkit
[INFO ] - idea.cycle.buffer.size: 1024
[INFO ] - sun.io.unicode.encoding: UnicodeBig
[INFO ] - javax.swing.rebaseCssSizeMap: true
[INFO ] - java.awt.printerjob: sun.lwawt.macosx.CPrinterJob
[INFO ] - idea.max.content.load.filesize: 20000
[INFO ] - com.jetbrains.suppressWindowRaise: false
[INFO ] - user.variant: 
[INFO ] - idea.no.launcher: false
[INFO ] - idea.smooth.progress: false
[INFO ] - java.vm.version: 25.265-b01
[INFO ] - path.separator: :
[INFO ] - jdk.http.auth.tunneling.disabledSchemes: ""
[INFO ] - os.version: 10.15.7
[INFO ] - java.vm.info: mixed mode
[INFO ] - java.home: /Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home/jre
[INFO ] - user.country: CN
[INFO ] - java.class.version: 52.0
[INFO ] - file.separator: /
[INFO ] - java.specification.name: Java Platform API Specification
[INFO ] - sun.cpu.isalist: 
[INFO ] - sun.jnu.encoding: UTF-8
[INFO ] - file.encoding.pkg: sun.io
[INFO ] - jna.tmpdir: /Users/hs/Library/Caches/JetBrains/IdeaIC2020.2/tmp
[INFO ] - java.endorsed.dirs: 
[INFO ] - java.util.concurrent.ForkJoinPool.common.threadFactory: com.intellij.concurrency.IdeaForkJoinWorkerThreadFactory
[INFO ] - sun.arch.data.model: 64
[INFO ] - user.timezone: Asia/Shanghai
[INFO ] - swing.bufferPerWindow: true
[INFO ] - java.version: 1.8.0_265
[INFO ] - CVS_PASSFILE: ~/.cvspass
[INFO ] - java.vm.compressedOopsMode: Zero based
[INFO ] - jna.nosys: true
[INFO ] - idea.fatal.error.notification: disabled
[INFO ] - io.netty.allocator.useCacheForAllThreads: false
[INFO ] - sun.boot.library.path: /Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home/jre/lib:/Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home/jre/lib
[INFO ] - jdk.attach.allowAttachSelf: true
[INFO ] - idea.version: 2020.2.3
[INFO ] - user.language: en
[INFO ] - idea.home.path: /Applications/IntelliJ IDEA CE.app/Contents
[INFO ] - idea.active: true
[INFO ] - java.rmi.server.disableHttp: true
[INFO ] - io.netty.allocator.cacheTrimIntervalMillis: 600000
[INFO ] - user.name: hs
[INFO ] - io.netty.serviceThreadPrefix: Netty 
[INFO ] - jna.loaded: true
[INFO ] - jb.vmOptionsFile: /Users/hs/Library/Application Support/JetBrains/IdeaIC2020.2/idea.vmoptions
[INFO ] - java.runtime.version: 1.8.0_265-b01
[INFO ] - java.vm.specification.vendor: Oracle Corporation
[INFO ] - sun.boot.class.path: /Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home/jre/lib/resources.jar:/Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home/jre/lib/rt.jar:/Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home/jre/lib/sunrsasign.jar:/Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home/jre/lib/jsse.jar:/Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home/jre/lib/jce.jar:/Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home/jre/lib/charsets.jar:/Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home/jre/lib/jfr.jar:/Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home/jre/classes
[INFO ] - os.name: Mac OS X
[INFO ] - jnidispatch.path: /Users/hs/Library/Caches/JetBrains/IdeaIC2020.2/tmp/jna13913931625936226942.tmp
[INFO ] - idea.vendor.name: JetBrains
[INFO ] - ai.djl.logging.level: debug
[INFO ] - idea.paths.selector: IdeaIC2020.2
[INFO ] - sun.management.compiler: HotSpot 64-Bit Tiered Compilers
[INFO ] - idea.executable: idea
[INFO ] - sun.java.launcher: SUN_STANDARD
[INFO ] - com.apple.mrj.application.live-resize: false
[INFO ] - java.vendor: AdoptOpenJDK
[INFO ] - 
[INFO ] - ----------Environment Variables----------
[INFO ] - PATH: /usr/local/Caskroom/miniconda/base/condabin:/Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home/bin:/Applications/IntelliJ IDEA CE.app/Contents/plugins/maven/lib/maven3/bin:/usr/local/mysql/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin
[INFO ] - SHELL: /bin/zsh
[INFO ] - CONDA_EXE: /usr/local/Caskroom/miniconda/base/bin/conda
[INFO ] - JAVA_HOME: /Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home
[INFO ] - HOMEBREW_BOTTLE_DOMAIN: https://mirrors.aliyun.com/homebrew/homebrew-bottles
[INFO ] - OLDPWD: /
[INFO ] - CONDA_PYTHON_EXE: /usr/local/Caskroom/miniconda/base/bin/python
[INFO ] - USER: hs
[INFO ] - TMPDIR: /var/folders/3g/2drk3qmj39b7h61rwf6w8f8w0000gn/T/
[INFO ] - SSH_AUTH_SOCK: /private/tmp/com.apple.launchd.SyN0JB68If/Listeners
[INFO ] - _CE_CONDA: 
[INFO ] - MAVEN_HOME: /Applications/IntelliJ IDEA CE.app/Contents/plugins/maven/lib/maven3
[INFO ] - XPC_FLAGS: 0x0
[INFO ] - MYSQL_HOME: /usr/local/mysql
[INFO ] - VERSIONER_PYTHON_VERSION: 2.7
[INFO ] - _CE_M: 
[INFO ] - JAVA_MAIN_CLASS_30188: ai.djl.integration.util.DebugEnvironment
[INFO ] - __CF_USER_TEXT_ENCODING: 0x1F5:0x0:0x0
[INFO ] - LOGNAME: hs
[INFO ] - LC_CTYPE: en_US.UTF-8
[INFO ] - XPC_SERVICE_NAME: com.jetbrains.intellij.ce.9060
[INFO ] - CONDA_SHLVL: 0
[INFO ] - HOME: /Users/hs
[INFO ] - 
[INFO ] - ----------Default Engine----------
[DEBUG] - cudart library not found.
[DEBUG] - Using cache dir: /Users/hs/.djl.ai/mxnet
[DEBUG] - Loading mxnet library from: /Users/hs/.djl.ai/mxnet/1.7.0-backport-mkl-osx-x86_64/libmxnet.dylib
[DEBUG] - Engine loaded from provider: MXNet
[DEBUG] - Found default engine: MXNet
[INFO ] - Engine name: MXNet
[INFO ] - Engine version: 1.7.0
[DEBUG] - Using cache dir: /Users/hs/.djl.ai/mxnet
[INFO ] - MXNet Library: /Users/hs/.djl.ai/mxnet/1.7.0-backport-mkl-osx-x86_64/libmxnet.dylib
[INFO ] - MXNet Features: BLAS_APPLE, CPU_SSE, SIGNAL_HANDLER, LAPACK, CPU_SSE2, CPU_SSE3, CPU_SSE4_1, OPENCV, MKLDNN
[INFO ] - 
[INFO ] - ----------Hardware----------
[INFO ] - Available processors (cores): 8
[INFO ] - Byte Order: LITTLE_ENDIAN
[INFO ] - Free memory (bytes): 206323872
[INFO ] - Maximum memory (bytes): 3817865216
[INFO ] - Total memory available to JVM (bytes): 257425408
[INFO ] - Heap committed: 257425408
[INFO ] - Heap nonCommitted: 20709376
[INFO ] - GPU Count: 0
[INFO ] - Default Device: cpu()
[INFO ] - GCC: 
[INFO ] - Apple clang version 12.0.0 (clang-1200.0.32.2)
[INFO ] - Target: x86_64-apple-darwin19.6.0
[INFO ] - Thread model: posix
[INFO ] - InstalledDir: /Library/Developer/CommandLineTools/usr/bin

--- a/api/src/main/java/ai/djl/training/dataset/DataIterable.java
+++ b/api/src/main/java/ai/djl/training/dataset/DataIterable.java
@@ -201,11 +201,10 @@ public class DataIterable implements Iterable<Batch>, Iterator<Batch> {
     }
 
     private void preFetch() {
-        List<Long> indices;
         if (!sample.hasNext()) {
             return;
         }
-        indices = sample.next();
+        List<Long> indices = sample.next();
         Callable<Batch> task = new PreFetchCallable(indices);
         Future<Batch> result = executor.submit(task);
         queue.offer(result);

--- a/basicdataset/src/main/java/ai/djl/basicdataset/TextDataset.java
+++ b/basicdataset/src/main/java/ai/djl/basicdataset/TextDataset.java
@@ -24,6 +24,8 @@ import ai.djl.ndarray.NDManager;
 import ai.djl.repository.Repository;
 import ai.djl.repository.Resource;
 import ai.djl.training.dataset.RandomAccessDataset;
+import java.util.ArrayList;
+import java.util.Comparator;
 import java.util.List;
 
 /**
@@ -45,6 +47,7 @@ public abstract class TextDataset extends RandomAccessDataset {
 
     protected Resource resource;
     protected boolean prepared;
+    protected List<Sample> samples;
 
     /**
      * Creates a new instance of {@link RandomAccessDataset} with the given necessary
@@ -113,6 +116,23 @@ public abstract class TextDataset extends RandomAccessDataset {
     }
 
     /**
+     * Returns a list of sample information.
+     *
+     * @return a list of sample information
+     */
+    public List<Sample> getSamples() {
+        if (samples == null) {
+            samples = new ArrayList<>();
+            for (int i = 0; i < size(); i++) {
+                List<String> text = getProcessedText(i, true);
+                samples.add(new Sample(i, text.size()));
+            }
+            samples.sort(Comparator.comparingInt(o -> o.sentenceLength));
+        }
+        return samples;
+    }
+
+    /**
      * Performs pre-processing steps on text data such as tokenising, applying {@link
      * ai.djl.modality.nlp.preprocess.TextProcessor}s, creating vocabulary, and word embeddings.
      *
@@ -124,6 +144,42 @@ public abstract class TextDataset extends RandomAccessDataset {
         TextData textData = source ? sourceTextData : targetTextData;
         textData.preprocess(
                 manager, newTextData.subList(0, (int) Math.min(limit, newTextData.size())));
+    }
+
+    /** A class stores {@code TextDataset} sample information. */
+    public static final class Sample {
+
+        private int sentenceLength;
+        private long index;
+
+        /**
+         * Constructs a new {@code Sample} instance.
+         *
+         * @param index the index
+         * @param sentenceLength the sentence length
+         */
+        public Sample(int index, int sentenceLength) {
+            this.index = index;
+            this.sentenceLength = sentenceLength;
+        }
+
+        /**
+         * Returns the sentence length.
+         *
+         * @return the sentence length
+         */
+        public int getSentenceLength() {
+            return sentenceLength;
+        }
+
+        /**
+         * Returns the sample index.
+         *
+         * @return the sample index
+         */
+        public long getIndex() {
+            return index;
+        }
     }
 
     /** Abstract Builder that helps build a {@link TextDataset}. */

--- a/basicdataset/src/main/java/ai/djl/basicdataset/utils/FixedBucketSampler.java
+++ b/basicdataset/src/main/java/ai/djl/basicdataset/utils/FixedBucketSampler.java
@@ -17,12 +17,13 @@ import ai.djl.training.dataset.RandomAccessDataset;
 import ai.djl.training.dataset.Sampler;
 import ai.djl.util.RandomUtils;
 import java.util.ArrayList;
-import java.util.Comparator;
+import java.util.Collections;
 import java.util.HashSet;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Set;
-import java.util.TreeSet;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /**
  * {@code FixedBucketSampler} is a {@code Sampler} to be used with {@link TextDataset}, and {@link
@@ -31,10 +32,11 @@ import java.util.TreeSet;
  * sampling is random across epochs.
  */
 public class FixedBucketSampler implements Sampler {
-    private Set<Bucket> buckets;
+
+    private static final Logger logger = LoggerFactory.getLogger(FixedBucketSampler.class);
+
     private int numBuckets;
     private int batchSize;
-    private boolean dropLast;
     private boolean shuffle;
 
     /**
@@ -43,14 +45,15 @@ public class FixedBucketSampler implements Sampler {
      *
      * @param batchSize the batch size
      * @param numBuckets the number of buckets
-     * @param dropLast whether to drop the last incomplete batch
      * @param shuffle whether to shuffle data randomyl while sampling
      */
-    public FixedBucketSampler(int batchSize, int numBuckets, boolean dropLast, boolean shuffle) {
+    public FixedBucketSampler(int batchSize, int numBuckets, boolean shuffle) {
         this.numBuckets = numBuckets;
         this.batchSize = batchSize;
-        this.dropLast = dropLast;
         this.shuffle = shuffle;
+        if (batchSize == 1) {
+            logger.warn("FixedBucketSampler is not meaningful with batch size 1.");
+        }
     }
 
     /**
@@ -61,7 +64,7 @@ public class FixedBucketSampler implements Sampler {
      * @param numBuckets the number of buckets
      */
     public FixedBucketSampler(int batchSize, int numBuckets) {
-        this(numBuckets, batchSize, false, true);
+        this(batchSize, numBuckets, true);
     }
 
     /**
@@ -71,13 +74,17 @@ public class FixedBucketSampler implements Sampler {
      * @param batchSize the batch size
      */
     public FixedBucketSampler(int batchSize) {
-        this(10, batchSize);
+        this(batchSize, 10);
     }
 
     /** {@inheritDoc} */
     @Override
     public Iterator<List<Long>> sample(RandomAccessDataset dataset) {
-        return new Iterate(dataset);
+        if (!(dataset instanceof TextDataset)) {
+            throw new IllegalArgumentException(
+                    "FixedBucketSampler can only be used with TextDataset");
+        }
+        return new Iterate((TextDataset) dataset);
     }
 
     /** {@inheritDoc} */
@@ -86,119 +93,70 @@ public class FixedBucketSampler implements Sampler {
         return batchSize;
     }
 
-    private static class Sample {
-        int sentenceLength;
-        long index;
-
-        public Sample(int index, int sentenceLength) {
-            this.index = index;
-            this.sentenceLength = sentenceLength;
-        }
-    }
-
-    private static class Bucket {
-        Set<Sample> samples;
-        int index;
-
-        public Bucket(int index, Set<Sample> samples) {
-            this.index = index;
-            this.samples = samples;
-        }
-    }
-
     private class Iterate implements Iterator<List<Long>> {
-        private long current;
-        private long size;
 
-        public Iterate(RandomAccessDataset dataset) {
-            if (dropLast) {
-                this.size = dataset.size() / batchSize;
-            } else {
-                this.size = (dataset.size() + batchSize - 1) / batchSize;
-            }
+        private List<List<TextDataset.Sample>> buckets;
+        private List<int[]> bucketBatch;
+        private int current;
 
-            if (!(dataset instanceof TextDataset)) {
-                throw new IllegalStateException(
-                        "FixedBucketSampler can only be used with TextDataset");
+        public Iterate(TextDataset dataset) {
+            buckets = new ArrayList<>(numBuckets);
+            bucketBatch = new ArrayList<>();
+            List<TextDataset.Sample> samples = dataset.getSamples();
+            int min = samples.get(0).getSentenceLength();
+            int max = samples.get(samples.size() - 1).getSentenceLength();
+            int step = Math.max((1 + max - min) / numBuckets, 1);
+            Set<Integer> set = new HashSet<>(numBuckets);
+            for (int i = 0; i < numBuckets; ++i) {
+                set.add(Math.max(max - (numBuckets - i - 1) * step, min));
             }
-            if (buckets == null) {
-                List<Sample> samples = new ArrayList<>();
-                for (int i = 0; i < dataset.size(); i++) {
-                    samples.add(
-                            new Sample(
-                                    i, ((TextDataset) dataset).getProcessedText(i, true).size()));
-                }
-                samples.sort(Comparator.comparingInt(o -> o.sentenceLength));
-                buckets = new TreeSet<>(Comparator.comparingInt(o -> o.index));
-                int bucketSize = samples.size() / numBuckets;
-                int bucketNumber = 0;
-                for (int i = 0; i < samples.size(); i = i + bucketSize) {
-                    int end = i + bucketSize;
-                    if (end > samples.size()) {
-                        end = samples.size();
+            int[] bucketKeys = set.stream().mapToInt(Integer::intValue).toArray();
+
+            int index = 0;
+            List<TextDataset.Sample> list = new ArrayList<>();
+            for (TextDataset.Sample sample : samples) {
+                if (sample.getSentenceLength() > bucketKeys[index]) {
+                    if (!list.isEmpty()) {
+                        buckets.add(list);
+                        list = new ArrayList<>();
                     }
-                    buckets.add(new Bucket(bucketNumber++, new HashSet<>(samples.subList(i, end))));
+                    ++index;
                 }
+                list.add(sample);
+            }
+            if (!list.isEmpty()) {
+                buckets.add(list);
+            }
+            for (int i = 0; i < buckets.size(); ++i) {
+                List<TextDataset.Sample> bucket = buckets.get(i);
+                for (int j = 0; j < bucket.size(); j += batchSize) {
+                    bucketBatch.add(new int[] {i, j});
+                }
+            }
+            if (shuffle) {
+                Collections.shuffle(bucketBatch, RandomUtils.RANDOM);
+                buckets.forEach(l -> Collections.shuffle(l, RandomUtils.RANDOM));
             }
         }
 
         /** {@inheritDoc} */
         @Override
         public boolean hasNext() {
-            return current < size;
+            return current < bucketBatch.size();
         }
 
         /** {@inheritDoc} */
         @Override
         public List<Long> next() {
-            int collected = 0;
-            List<Sample> allSamples = new ArrayList<>();
-
-            Iterator<Bucket> iterator = buckets.iterator();
-            Bucket bucket = firstBucket(iterator);
-            while (collected < batchSize) {
-                Set<Sample> samples = bucket.samples;
-                List<Sample> bucketSamples = new ArrayList<>();
-                for (Sample sample : samples) {
-                    bucketSamples.add(sample);
-                    collected++;
-                    if (collected >= batchSize) {
-                        break;
-                    }
-                }
-                for (Sample sample : bucketSamples) {
-                    samples.remove(sample);
-                }
-                allSamples.addAll(bucketSamples);
-                if (collected >= batchSize) {
-                    break;
-                }
-                if (!iterator.hasNext()) {
-                    if (shuffle) {
-                        iterator = buckets.iterator();
-                    } else {
-                        throw new IllegalStateException("Code should never reach here");
-                    }
-                }
-                bucket = iterator.next();
-            }
-            List<Long> next = new ArrayList<>();
-            for (Sample sample : allSamples) {
-                next.add(sample.index);
+            int[] batch = bucketBatch.get(current);
+            List<Long> ret = new ArrayList<>();
+            List<TextDataset.Sample> bucket = buckets.get(batch[0]);
+            int end = Math.min(bucket.size(), batch[1] + batchSize);
+            for (int i = batch[1]; i < end; ++i) {
+                ret.add(bucket.get(i).getIndex());
             }
             current++;
-            return next;
-        }
-
-        private Bucket firstBucket(Iterator<Bucket> iterator) {
-            if (shuffle) {
-                int firstIndex = RandomUtils.nextInt(buckets.size());
-                for (int i = 0; i < firstIndex - 1; i++) {
-                    iterator.next();
-                }
-                return iterator.next();
-            }
-            return iterator.next();
+            return ret;
         }
     }
 }

--- a/basicdataset/src/test/java/ai/djl/basicdataset/FixedBucketSamplerTest.java
+++ b/basicdataset/src/test/java/ai/djl/basicdataset/FixedBucketSamplerTest.java
@@ -25,9 +25,10 @@ import org.testng.Assert;
 import org.testng.annotations.Test;
 
 public class FixedBucketSamplerTest {
+
     @Test
     public void testFixedBucketSampler() throws IOException, TranslateException {
-        FixedBucketSampler fixedBucketSampler = new FixedBucketSampler(10, 100, false, true);
+        FixedBucketSampler fixedBucketSampler = new FixedBucketSampler(10, 10, false);
         TatoebaEnglishFrenchDataset dataset =
                 TatoebaEnglishFrenchDataset.builder()
                         .setSampling(fixedBucketSampler)
@@ -41,47 +42,28 @@ public class FixedBucketSamplerTest {
                                         .optIncludeValidLengths(true)
                                         .addPad(0, 0, (m) -> m.ones(new Shape(1)), 10)
                                         .build())
+                        .optLimit(200)
                         .build();
+
+        dataset.prepare();
 
         Iterator<List<Long>> iterator = fixedBucketSampler.sample(dataset);
         long count = 0;
         Set<Long> indicesSet = new HashSet<>();
         while (iterator.hasNext()) {
             List<Long> indices = iterator.next();
-            int max = Integer.MIN_VALUE;
-            int min = Integer.MAX_VALUE;
-            for (Long index : indices) {
-                int size = dataset.getProcessedText(index, true).size();
-                if (size > max) {
-                    max = size;
-                }
-                if (size < min) {
-                    min = size;
-                }
-            }
             indicesSet.addAll(indices);
-            count = count + indices.size();
+            count += indices.size();
         }
         Assert.assertEquals(count, dataset.size());
         Assert.assertEquals(indicesSet.size(), dataset.size());
 
-        fixedBucketSampler = new FixedBucketSampler(10, 100, false, false);
+        fixedBucketSampler = new FixedBucketSampler(10, 5, true);
         iterator = fixedBucketSampler.sample(dataset);
         count = 0;
-        indicesSet = new HashSet<>();
+        indicesSet.clear();
         while (iterator.hasNext()) {
             List<Long> indices = iterator.next();
-            int max = Integer.MIN_VALUE;
-            int min = Integer.MAX_VALUE;
-            for (Long index : indices) {
-                int size = dataset.getProcessedText(index, true).size();
-                if (size > max) {
-                    max = size;
-                }
-                if (size < min) {
-                    min = size;
-                }
-            }
             indicesSet.addAll(indices);
             count = count + indices.size();
         }

